## Importing necessary modules

In [1]:
import time

start_time = time.time()

In [2]:
from sklearnex import patch_sklearn
patch_sklearn()

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder,MinMaxScaler
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error,mean_absolute_error
from sklearn.model_selection import RepeatedKFold,cross_val_score
from sklearn.model_selection import GridSearchCV
import warnings
warnings.filterwarnings("ignore")

Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)


## Loading data and exploration

In [3]:
estate=pd.read_csv(r"K:\Desktop\NIIT\tables\DS2_C5_S7_KCHouse_Data_Practice.csv")
estate.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

## Data preprocessing

In [4]:
estate.describe().T

count          mean           std           min  \
id             21613.0  4.580302e+09  2.876566e+09  1.000102e+06   
price          21613.0  5.400881e+05  3.671272e+05  7.500000e+04   
bedrooms       21613.0  3.370842e+00  9.300618e-01  0.000000e+00   
bathrooms      21613.0  2.114757e+00  7.701632e-01  0.000000e+00   
sqft_living    21613.0  2.079900e+03  9.184409e+02  2.900000e+02   
sqft_lot       21613.0  1.510697e+04  4.142051e+04  5.200000e+02   
floors         21613.0  1.494309e+00  5.399889e-01  1.000000e+00   
waterfront     21613.0  7.541757e-03  8.651720e-02  0.000000e+00   
view           21613.0  2.343034e-01  7.663176e-01  0.000000e+00   
condition      21613.0  3.409430e+00  6.507430e-01  1.000000e+00   
grade          21613.0  7.656873e+00  1.175459e+00  1.000000e+00   
sqft_above     21613.0  1.788391e+03  8.280910e+02  2.900000e+02   
sqft_basement  21613.0  2.915090e+02  4.425750e+02  0.000000e+00   
yr_built       21613.0  1.971005e+03  2.937341e+01  1.900000e+03   
yr_renovated   21613.0  8.440226e+01  4.016792e+02  0.000000e+00   
zipcode        21613.0  9.807794e+04  5.350503e+01  9.800100e+04   
lat            21613.0  4.756005e+01  1.385637e-01  4.715590e+01   
long           21613.0 -1.222139e+02  1.408283e-01 -1.225190e+02   
sqft_living15  21613.0  1.986552e+03  6.853913e+02  3.990000e+02   
sqft_lot15     21613.0  1.276846e+04  2.730418e+04  6.510000e+02   

                        25%           50%           75%           max  
id             2.123049e+09  3.904930e+09  7.308900e+09  9.900000e+09  
price          3.219500e+05  4.500000e+05  6.450000e+05  7.700000e+06  
bedrooms       3.000000e+00  3.000000e+00  4.000000e+00  3.300000e+01  
bathrooms      1.750000e+00  2.250000e+00  2.500000e+00  8.000000e+00  
sqft_living    1.427000e+03  1.910000e+03  2.550000e+03  1.354000e+04  
sqft_lot       5.040000e+03  7.618000e+03  1.068800e+04  1.651359e+06  
floors         1.000000e+00  1.500000e+00  2.000000e+00  3.500000e+00  
waterfront     0.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00  
view           0.000000e+00  0.000000e+00  0.000000e+00  4.000000e+00  
condition      3.000000e+00  3.000000e+00  4.000000e+00  5.000000e+00  
grade          7.000000e+00  7.000000e+00  8.000000e+00  1.300000e+01  
sqft_above     1.190000e+03  1.560000e+03  2.210000e+03  9.410000e+03  
sqft_basement  0.000000e+00  0.000000e+00  5.600000e+02  4.820000e+03  
yr_built       1.951000e+03  1.975000e+03  1.997000e+03  2.015000e+03  
yr_renovated   0.000000e+00  0.000000e+00  0.000000e+00  2.015000e+03  
zipcode        9.803300e+04  9.806500e+04  9.811800e+04  9.819900e+04  
lat            4.747100e+01  4.757180e+01  4.767800e+01  4.777760e+01  
long          -1.223280e+02 -1.222300e+02 -1.221250e+02 -1.213150e+02  
sqft_living15  1.490000e+03  1.840000e+03  2.360000e+03  6.210000e+03  
sqft_lot15     5.100000e+03  7.620000e+03  1.008300e+04  8.712000e+05

In [5]:
estate.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [6]:
estate.isna().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [7]:
estate.shape

(21613, 21)

LabelEncoding , Scaling and Train_test_split

In [8]:
scaler=MinMaxScaler()
encoder=LabelEncoder()
x=estate.loc[:,(estate.columns!="price") & (estate.columns!="id")]
y=estate.price

## encoding
x["date"]=encoder.fit_transform(x.date)

cv = RepeatedKFold(n_splits=10, n_repeats=3, random_state=1)

In [9]:
all_scores,names=[],[]

## Model creation

In [10]:
lr=LinearRegression()

scores = cross_val_score(lr, x, y, scoring='neg_mean_squared_error', cv=cv, n_jobs=-1, error_score='raise')

print("Training Score of model : ",np.mean(np.sqrt(-scores)))

names.append('Linear')
all_scores.append(np.mean(np.sqrt(-scores)))

Training Score of model :  201211.12683122358


### Bagging

In [11]:
from sklearn.ensemble import RandomForestRegressor,AdaBoostRegressor

In [12]:
rf_model=RandomForestRegressor(max_depth=7)


scores = cross_val_score(rf_model, x, y, scoring='neg_mean_squared_error', cv=cv, n_jobs=-1, error_score='raise')


print("Training Score of model : ",np.mean(np.sqrt(-scores)))

names.append('RandomForest')

all_scores.append(np.mean(np.sqrt(-scores)))

Training Score of model :  151094.7029422585


### AdaBoosting

In [13]:
# evaluate the model
booster=AdaBoostRegressor(base_estimator=DecisionTreeRegressor(max_depth=7),n_estimators=50)

scores = cross_val_score(booster, x, y, scoring='neg_mean_squared_error', cv=cv, n_jobs=-1, error_score='raise')

print("Training Score of model : ",np.mean(np.sqrt(-scores)))

names.append('AdaBooster')
all_scores.append(np.mean(np.sqrt(-scores)))

Training Score of model :  150520.01848250846


## Stacking

In [14]:
# required Python libraries
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
from sklearn.ensemble import StackingRegressor

In [15]:
def get_stacking():
    model1=[]
    model1.append(("Linear",LinearRegression()))
    model1.append(("Decision",DecisionTreeRegressor(max_depth=7)))
    model1.append(("rf_Regressor",RandomForestRegressor(max_depth=7)))
    #model1.append(("SVR",SVR()))
    level1 = LinearRegression()
    model = StackingRegressor(estimators=model1, final_estimator=level1, cv=3)
    return model

stack_model=get_stacking()
scores = cross_val_score(stack_model,x,y, scoring='neg_mean_squared_error', cv=cv, n_jobs=-1, error_score='raise')
all_scores.append(np.mean(np.sqrt(-scores)))
names.append('stacking')

### Final Verdict from results

In [16]:
for i in range(len(names)):
    print("{0} score : {1}".format(names[i],all_scores[i]))

Linear score : 201211.12683122358
RandomForest score : 151094.7029422585
AdaBooster score : 150520.01848250846
stacking score : 147835.64837576257


## Conclusion:
The Stacking model has the lowest RMSE score of 147835.65, making it the best performing model in this comparison. The AdaBoost model has a score of 150520.02, followed by the Random Forest model with a score of 151094.70. The Linear model has the highest RMSE score of 201211.13.

In conclusion, the Stacking model is the best performing model in this comparison, followed by the AdaBoost model and then the Random Forest model. Further exploration and optimization may be necessary to determine the optimal model for the given data set.

In [17]:
end_time = time.time()

print("Time taken: ", end_time - start_time, "seconds")

Time taken:  92.81975746154785 seconds
